### ISSUE TITLE
New `Should Contain Any` and `Should Not Contain Any` keywords

### ISSUE THREAD

**Issue by jongu82** [robotframework/robotframework#2120]
I am searching through a list and would like to pass if any one of my values are found.  It appears that the List Should Contain Value keyword only accepts one value.  I have tried using Value1 or Value2 but it does not appear to be searching for either one.  I would suggest to modify this keyword to accept multiple values separated by 'or'.

Furthermore, I would like to have the output display the matches, and fail if no matches are found

**Comment by pekkaklarck** [comment]
Yes, there doesn't seem to be any keyword that supports this in Collections. We cannot change `List Should Contain Value` because basically any change would be backwards incompatible. Adding separate `List Should Contain Any Value` is possible, but we could also consider adding more generic `Should Contain Any` to BuiltIn. Are you interested to provide a pull request?

**Comment by jongu82** [comment]
Yes I am interested.  I'm fairly new to github and the use of pull requests but I'd be willing to give it a shot.

**Comment by pekkaklarck** [comment]
GitHub has pretty good documentation about pull requests, and our contributing guidelines have links to them along with some other valuable information.

Would you be interested to implement `List Should Contain Any Value` or `Should Contain Any`? I think the latter would be more valuable because it would be more generic. If the former is implemented, then we should add related dict keywords too. Adding all of them is possible too but perhaps too much work compared to benefits. Regardless what is implemented, `Not` version (e.g. `Should Not Contain Any`) should be added too.

**Comment by pekkaklarck** [comment]
After thinking this a bit more, I believe it's best to implement `Should Contain Any` and `Should Not Contain Any`. They are generic and can be used with strings or anything that is list-like. This functionality isn't needed too often, so adding separate `List Should Contain Any Value`, `Dictionary Should Contain Any Key`, etc. wouldn't be worth the effort in my opinion.

Anyone interested to implement `Should (Not) Contain Any`? There's plenty of time until RF 3.0 which is the next planned release and I can help if needed. If you are new to providing pull requests and contributing in general, you can start by reading these guidelines:
https://github.com/robotframework/robotframework/blob/master/CONTRIBUTING.rst

**Comment by pekkaklarck** [comment]
Nobody has shown any interest to implement the new keywords, but I consider them so useful that I plan to include them in RF 3.0.1 anyway. If someone is interested in contributing to RF, this would be a pretty easy issue to get started.

**Comment by jimni** [comment]
I would like to have a go at this issue. 
I will either come up with the solution before 7th of June, or consider me forfeit :)

**Comment by pekkaklarck** [comment]
Great! Please provide a pull request once you got something ready and we can review it. If you do it in a separate branch, you can easily commit new changes  and the PR automatically gets updated.

The signature of these keywords should probably something like `container, *items`. That allows usage like this:

    Should Contain Any    ${text}    foo    bar    zap

What do you think, should same `msg=None` and `values=True` arguments be accepted as `Should (Not) Contain` accepts? If yes, the signature should probably be `container, *items, **config*` and code contain something like:

```python
    msg = config.pop('msg', None)
    values = config.pop('values', True)
    if config:
        raise RuntimeError('Message explaining that unsupported config values were used')
```

Alternatively we could look does the last and second last item start with `msg=` and `values=` but that is more complicated. This kind of `**config` approach is already used by the Process library and it has worked well there. If we didn't need to care about Python 2, we could just use [keyword-only arguments](https://www.python.org/dev/peps/pep-3102/).

**Comment by jimni** [comment]
> What do you think, should same `msg=None` and `values=True` arguments be accepted as `Should (Not) Contain` accepts?
I believe they should. I will aim for this behaviour.

**Comment by jimni** [comment]
Should we also redefine `Should (Not) Contain` as wrappers for these new keywords? As in:
    def should_not_contain(self, container, item, msg=None, values=True):
        self.should_not_contain_any(container, [item], msg, values)

I'm not sure if this interpretation of SRP is considered appropriate in RF, please let me know.

**Comment by pekkaklarck** [comment]
Depends how the code would look line in the end. These kws may also have slightly different failure messages making direct reuse harder.

**Comment by jimni** [comment]
I see your point. Let's consider that afterwards

**Comment by jimni** [comment]
While writing tests I came across the fact that lists and tuples are stringified with `u' '` (unlike strings):

    '(u'yet', u'yes', u'no')'

I'm not sure if this is desired behaviour or not, or maybe I'm just doing something wrong.
For now, I left this hanging in the tests' fail messages, because it looks harmless. Let me know if this needs a follow-up resolution. 

Looking forward to your feedback.

**Comment by pekkaklarck** [comment]
Is this in the error message? Probably should use `utils.seq2str` there.

`u` prefix is added because strings passed to Robot are Unicode and string representation of tuples and lists contains the `repr()` of items. It's not added in Python 3 which is problematic in tests. Using the aforementioned util resolves that too.

**Comment by pekkaklarck** [comment]
On a very quick look the PR looks very good. Hopefully I have time to look it tomorrow in more detail. Trying to get our house sold means I'm somewhat busy.

**Comment by jimni** [comment]
Take your time, I'm not going anywhere.

**Comment by jimni** [comment]
`utils.seq2str` worked very well. Thank you.

**Comment by pekkaklarck** [comment]
This has been implemented by @jimni in PR #2393 that was just merged. I'll do a little additional testing possibly little fine tuning but otherwise this issue can be considered done. Thanks @jimni for a great PR!

**Comment by pekkaklarck** [comment]
Final fine-tuning in done in 5bde125. Main change was making it an explicit error to use these keywords without any items to test against. `Should Contain Any` already failed with a bad error and `Should Not Contain Any` passed in this case. I think it is safer to fail in both cases and to have a better message, but it is still easy to change this before the 3.0.1 release.

I consider this issue done. Already started thinking should we also add `Should (Not) Match Any` keywords... =)
